File: app/src/main/java/com/github/lany192/mnist/KerasTFLite.java
package com.github.lany192.mnist;

import android.content.Context;
import android.content.res.AssetManager;

import org.tensorflow.lite.Interpreter;

import java.io.BufferedOutputStream;
import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.util.ArrayList;
import java.util.List;

public class KerasTFLite {
    private static final String MODEL_FILE = "model.tflite";
    private Interpreter mInterpreter;

    public KerasTFLite(Context context) throws IOException {
        File file = loadModelFile(context);
        mInterpreter = new Interpreter(file);
    }

    public String run(float[] input) {
        //result will be number between 0~9
        float[][] labelProbArray = new float[1][10];
        mInterpreter.run(input, labelProbArray);
        List<String> labels = new ArrayList<>();
        for (int i = 0; i < 10; i++) {
            labels.add(String.valueOf(i));
        }
        return labels.get(getMax(labelProbArray[0]));
    }

    private File loadModelFile(Context context) throws IOException {
        String filePath = context.getFilesDir() + File.separator + MODEL_FILE;
        File file = new File(filePath);
        if (!file.exists()) {
            AssetManager assetManager = context.getAssets();
            InputStream stream = assetManager.open(MODEL_FILE);
            OutputStream output = new BufferedOutputStream(new FileOutputStream(filePath));
            byte[] buffer = new byte[1024];
            int read;
            while ((read = stream.read(buffer)) != -1) {
                output.write(buffer, 0, read);
            }
            stream.close();
            output.close();
        }
        return file;
    }

    private int getMax(float[] results) {
        int maxID = 0;
        float maxValue = results[maxID];
        for (int i = 1; i < results.length; i++) {
            if (results[i] > maxValue) {
                maxID = i;
                maxValue = results[maxID];
            }
        }
        return maxID;
    }

    public void release() {
        mInterpreter.close();
    }
}

File: app/src/main/kotlin/com/github/lany192/mnist/KerasTFLite.kt
package com.github.lany192.mnist

import android.content.Context
import android.os.Environment
import org.tensorflow.lite.Interpreter
import java.io.BufferedOutputStream
import java.io.File
import java.io.FileOutputStream
import java.io.IOException

class KerasTFLite(context: Context) {
    private val mInterpreter: Interpreter

    init {
        val file = loadModelFile(context)
        mInterpreter = Interpreter(file)
    }

    fun run(input: FloatArray?): String? {
        if (input == null || input.size != MODEL_SIZE * MODEL_SIZE) return null
        val input4d = Array(1) { Array(MODEL_SIZE) { Array(MODEL_SIZE) { FloatArray(1) } } }
        for (i in input.indices) {
            input4d[0][i / MODEL_SIZE][i % MODEL_SIZE][0] = input[i]
        }
        val output = Array(1) { FloatArray(10) }
        mInterpreter.run(input4d, output)
        return getMax(output[0]).toString()
    }

    @Throws(IOException::class)
    private fun loadModelFile(context: Context): File {
        val modelFile = "model.tflite"
        val filePath = context.filesDir.path + File.separator + modelFile
        val file = File(filePath)
        // 每次都从 assets 覆盖拷贝：filesDir 在 APK 更新后仍保留旧文件，
        // 若按存在性跳过拷贝，会一直加载到旧模型
        val assetManager = context.assets
        assetManager.open(modelFile).use { stream ->
            BufferedOutputStream(FileOutputStream(file)).use { output ->
                val buffer = ByteArray(8192)
                var read: Int
                while (stream.read(buffer).also { read = it } != -1) {
                    output.write(buffer, 0, read)
                }
            }
        }
        return file
    }

    private fun getMax(results: FloatArray): Int {
        var maxID = 0
        var maxValue = results[maxID]
        for (i in 1..<results.size) {
            if (results[i] > maxValue) {
                maxID = i
                maxValue = results[maxID]
            }
        }
        return maxID
    }

    fun release() {
        mInterpreter.close()
    }

    companion object {
        private const val MODEL_SIZE = 28
    }
}
